### PR TITLE
HOFF-617: Add trivy, Snyk, SonarQube and Cron Jobs for tearing down envs

### PR DESCRIPTION
**Pull request by adityababumallisettiHO**
What?

Trivy detects vulnerabilities of OS packages, Docker Images and application dependencies.  Snyk doesn't scan Docker Images but with webUI we can easily find out the vulnerabilities. It also scans the Kubernetes manifest files. SonarQube is a Code Quality Assurance tool that collects and analyzes source code, and provides reports for the code quality of your project.

This covers two tickets
[https://collaboration.homeoffice.gov.uk/jira/browse/HOFF-617](url)
[https://collaboration.homeoffice.gov.uk/jira/browse/TLF-30](url)

Why?
A security vulnerability is a software code flaw or a system misconfiguration through which attackers can directly gain unauthorized access to a system or network. Once inside, the attacker can leverage authorizations and privileges to compromise systems and assets.
An exploit is the means through which hackers use a vulnerability to mount an attack. An exploit is typically a piece of specially crafted software or a sequence of commands.
With SonarQube, your developers can ensure source code quality and application security by identifying and rectifying code duplications and potential bugs.

How?
Trivy, Snyk and Sonar Scanners steps added to the Pipelines. snyk_token added to Drone secrets. Added Snyk package to package.json and recreated the yarn.lock file. Snyk account created and the token is used as Secret. Snyk secret is also added to Git webhook url.
Added Cron jobs for scanner and tear down branches in drone pipeline file
Added the Cron Job Schedule expressions in Drone crons.
Added Documentation in confluence
[https://collaboration.homeoffice.gov.uk/display/DSASS/Configure+Drone+CLI+to+create+Drone+Cron+Jobs](url) 

Link for Cron Jobs in Drone: [https://drone-gh.acp.homeoffice.gov.uk/UKHomeOffice/lmr/settings/cron](url)
Link for Drone Job [https://drone-gh.acp.homeoffice.gov.uk/UKHomeOffice/lmr/191/1/8](url)

Testing?
Screenshots (optional)
image
image
image

### CODE REVIEW

File: .drone.yml [hunk 1]
```diff
@@ -16,6 +16,7 @@ environment:
   HOF_CONFIG: hof-services-config/Landlords_Make_A_Report
   NON_PROD_AVAILABILITY: Mon-Sun 08:00-23:00 Europe/London
   READY_FOR_TEST_DELAY: 20s
+  # NOTIFY_STUB: stub
```
Comment (sulthan-ahmed) [inline .drone.yml:19]: suggest deleting this

File: .drone.yml [hunk 2]
```diff
@@ -209,6 +272,49 @@ steps:
...
+  #   environment:
+  #     DRONE_SERVER:
+  #       from_secret: drone_server
+  #     DRONE_TOKEN:
+  #       from_secret: drone_token
+  #   commands:
+  #     - bin/sanity_check_build.sh
+  #   when:
+  #     target: PROD
+  #     event: promote
```
Comment (sulthan-ahmed) [inline .drone.yml:303]: probably best to delete it tbh

File: .drone.yml [hunk 3]
```diff
@@ -266,6 +372,57 @@ steps:
...
+  #   environment:
+  #     DRONE_GIT_USERNAME:
+  #       from_secret: drone_git_username
+  #     DRONE_GIT_TOKEN:
+  #       from_secret: drone_git_token
+  #   commands:
+  #     - git clone https://$${DRONE_GIT_USERNAME}:$${DRONE_GIT_TOKEN}@github.com/UKHomeOfficeForms/hof-services-config.git
+  #   when:
+  #     cron: security_scans
+  #     event: cron
```
Comment (sulthan-ahmed) [inline .drone.yml:387]: prob delete tbh

File: .snyk [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      expires: '2024-11-01T17:02:21.865Z'
+  SNYK-JS-NODEMAILER-6219989:
+    - '*':
+      reason: To be updated to a newer version in hof
+      expires: '2024-11-01T17:02:21.865Z'
+  SNYK-JS-TOUGHCOOKIE-5672873:
+    - '*':
+      reason: To be updated to a newer version in hof
+      expires: '2024-11-01T17:02:21.865Z'
+patch: {}
```
Comment (sulthan-ahmed) [inline .snyk:57]: new line?

File: .drone.yml [hunk 5]
```diff
@@ -209,6 +272,49 @@ steps:
...
+      DRONE_GIT_USERNAME:
+        from_secret: drone_git_username
+      DRONE_GIT_TOKEN:
+        from_secret: drone_git_token
+    commands:
+      - git clone https://$${DRONE_GIT_USERNAME}:$${DRONE_GIT_TOKEN}@github.com/UKHomeOfficeForms/hof-services-config.git
+    when:
+      target: PROD
+      event: promote
+
```
Comment (sulthan-ahmed) [inline .drone.yml:317]: do you think we can simplify all of this for all the different environments, because it's just running the same steps

**Comment by adityababumallisettiHO** [comment]
Waiting for @mislam987 to get back on the commented Drone CI steps. I have requested @mislam987 to check with the team if we require these steps or remove them for now?

**Comment by sulthan-ahmed** [comment]
ok @adityababumallisettiHO

File: .drone.yml (hunk 1 above)
Comment (adityababumallisettiHO) [reply .drone.yml]: This has been deleted. I have also checked with Muhammed Islam and Rhodine

File: .drone.yml (hunk 3 above)
Comment (adityababumallisettiHO) [reply .drone.yml]: We have been using this step for rest of services. We need to run this cron step of cloning the repos to scan vulnerabilities with Snyk. I have uncommented this step.

File: .drone.yml (hunk 5 above)
Comment (adityababumallisettiHO) [reply .drone.yml:285]: I have looked into it. I believe it is better to leave it as it is. It is difficult to use this step for all the envs as the conditions are different for not-prod and prod.

**Review by jamiecarterHO — CHANGES_REQUESTED** [review]
Just one thing I found to update.

Other is that there is a new vuln in the pipeline for notifications-node-client. As its a new service perhaps notifications-node-client v8.0.0 could be used? I believe this vuln is fixed there.

File: sonar-project.properties [hunk 6]
```diff
@@ -0,0 +1,6 @@
+sonar.projectKey=Illegal-Migration-Act
+sonar.projectName=Illegal-Migration-Act
```
[the remark this answers is not in the document]
Comment (adityababumallisettiHO) [reply sonar-project.properties]: I have replaced them with Landlords-Make-A-Report

**Comment by adityababumallisettiHO** [comment]
Added notifications-node-client v8.0.0 resolves vulnerability in package.json

**Comment by adityababumallisettiHO** [comment]
There is a ticket temi is working on fixing these vulnerabilties
https://collaboration.homeoffice.gov.uk/jira/browse/HOFF-694

I have added a ticket for LMR service 
https://collaboration.homeoffice.gov.uk/jira/browse/HOFF-696
and for LCS which is cloned version of LMR service 
https://collaboration.homeoffice.gov.uk/jira/browse/HOFF-697

**Comment by jamiecarterHO** [comment]
@adityababumallisettiHO I suppose you could add the new vulnerabilities that can't be updated in the project to the snyk whitelist

**Comment by adityababumallisettiHO** [comment]
Hi @jamiecarterHO ,
Vulnerabilities have been added to snyk ignore file.
We have all checks passed.
Drone CI: https://drone-gh.acp.homeoffice.gov.uk/UKHomeOffice/lmr/270

image

**Review by jamiecarterHO — APPROVED** [review]
Looks good, I can see build is all green for branch pipeline and scanners running.
Make sure to squash those commits and check the commit notes :)

**Pull request merged**